software/17.Kubernetes/README.md

```diff
@@ -0,0 +1,159 @@
+
+```
+kubectl get all
+```
+
+**Resources**
+
+- `kubectl explain deployment.*`
+- [Deployment documentation](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/)
+
+## Step 3: At your service
+
+You might have noticed, that when you run `kubectl get all`, the pod associated to the `server-deployment` is in a state of `CrashLoopBackOff` and is said to have restarted multiple times.
+
+```
+...
+NAME                                    READY   STATUS             RESTARTS   AGE
+pod/server-deployment-5ddbdc478-qqcwr   0/1     CrashLoopBackOff   4          115s
+...
+```
+
+If we investigate the issue by inspecting the deployment's logs  (`kubectl logs deployment.apps/server-deployment`)  we get the following output.
+
+```
+panic: server selection error: context deadline exceeded, current topology: { Type: Unknown, Servers: [{ Addr: mongo-service:27017, Type: Unknown, Average RTT: 0, Last error: connection() error occured during connection handshake: dial tcp: lookup mongo-service on 10.96.0.10:53: server misbehaving }, ] }
+
+goroutine 1 [running]:
+gokube/models.init.0()
+	/app/models/db.go:26 +0x2d4
+```
+
+Weird! The server application is telling us it cannot connect to the Mongo database.
+
+That's because, for now, the `mongo-deployment` and the `server-deployment` are isolated from one another. They cannot comunicate.
+
+This is where **Kubernetes services** come into play. A service allows Kubernetes object to expose themselves to other objects.
+
+Your job is to create a service for the `mongo-deployment` so it's accessible by other objects.
+
+The service for the `mongo-deployment` must:
+
+- Be of kind `Service`.
+- Have the name `mongo-service`.
+- Forward the port `27017` to the target port `27017`.
+- Have the `spec.selector.app` set to `mongo`. This will ensure the service is linked to your `mongo-deployment`.
+
+Once it's done, apply all your changes and run `kubectl get all`. Your deployments should be running without any errors.
+
+You can inspect the `server-deployment`'s logs to make sure everything is okay.
+
+**Resources**
+
+- `kubectl explain service.*`
+- [Service documentation](https://kubernetes.io/docs/concepts/services-networking/service/)
+- [Connecting applications using services](https://kubernetes.io/docs/concepts/services-networking/connect-applications-service/)
+- [Kubernetes services](https://www.bmc.com/blogs/kubernetes-services/)
+
+## Step 4: Make it robust
+
+You never know what can happen with your application. Crashes happen all the time and as a developper you can't afford **downtime** or **data loss**.
+
+Kubernetes already shields you against your app going offline. It will identify crashes and restart your app so it's always available. However, it's your job to make sure your data persists!
+
+In this final step, your job is to create a persistent storage for your database in case it crashes.
```

Comment:
```suggestion
Kubernetes already shields you against your app going offline. It will identify crashes and restart your app so it's always available. However, it's your job to make sure your data persists!

In this final step, your job is to create a [persistent storage](https://kubernetes.io/docs/concepts/storage/persistent-volumes/) for your database in case it crashes.
```